src/back/hlsl/writer.rs

```diff
@@ -153,15 +153,13 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
         // Write all structs
         for (handle, ty) in module.types.iter() {
             if let TypeInner::Struct { ref members, span } = ty.inner {
-                if let Some(member) = members.last() {
-                    if let TypeInner::Array {
-                        size: crate::ArraySize::Dynamic,
-                        ..
-                    } = module.types[member.ty].inner
-                    {
-                        // unsized arrays can only be in storage buffers, for which we use `ByteAddressBuffer` anyway.
-                        continue;
-                    }
+                if module.types[members.last().unwrap().ty]
```

Comment:
An empty struct in input shader will break this code? I think it would be better to use `if let Some(member) = members.last() ....`

Reply:
Empty structures are forbidden in WGSL, and I thought the validator enforced this - but it does not. GLSL doesn't allow empty structures either, but SPIR-V does (and Vulkan doesn't seem to add any further restrictions about that).

But for the time being, since we do have empty structs passing through Naga, I'll change this and the other sites.

Reply:
@kvark suggested just banning empty structs in validation; that's implemented in #1826.

Reply:
Can we avoid unwrap here and return `EmptyStruct` error instead? In case where validation is disabled (it disabled by default).

Reply:
The general stance backends have over panicking is that it's fine if the IR is illegal there are multiple cases where we panic if the IR isn't good, just grep `unwrap`,  if validation isn't enabled it's the responsibility of the user to ensure it's legal, otherwise the behavior may include panics.

Reply:
That principle makes sense; I've submitted #1828 to write it down.